Support deletions in rearrange (#11814)

stefanvodita opened the pull request:
### Description

Changes made to support deletes in `IndexRearranger`:
1. Add `isDeleted(LeafReader reader, int idx)` in document selectors, which returns true if the document found at index `idx` in the leaf reader was marked for deletion in the original index.
2. Record all docs in the `DocumentSelector` (not just live ones).
3. When rearranging, in `addOneSegment`, add all docs that belong in the segment and then delete the ones that were originally marked for deletion.

### Testing
The new test class is copied from `TestIndexRearranger` with modifications for deleting documents in the original index and testing that they are still present and deleted in the rearranged index. In a subsequent revision, this file would be merged into `TestIndexRearranger`.

#11814

stefanvodita committed:
Support deletions in rearrange (#11814)

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/index/IndexRearranger.java:
```diff
@@ -175,5 +202,7 @@ public CacheHelper getReaderCacheHelper() {
   public interface DocumentSelector {
     BitSet getFilteredLiveDocs(CodecReader reader) throws IOException;
+
+    boolean isDeleted(LeafReader reader, int idx) throws IOException;
```
I think it is confusing to put this here:
1. The first method `getFilteredLiveDocs` is basically from "original" index, generating a set of documents that will be pulled into the "new" index. But the new method is basically from "new" index, query whether a document is deleted. You see that they basically are supposed to be applied to different index and it is quite confusing.
2. Given that this method are supposed to be applied to the "new" index, I would assume the creator will have some information on which documents are supposed to be deleted beforehand, like a set of ids that are deleted in the "origin" index. Then why don't we put this deletion logic after the rearrange has done? I don't see any advantage of putting the logic here. Also now the `writer.addIndexes` are subject to a lock which means we're slowing things down quite a lot.

stefanvodita replied:
1. Agreed, that can be confusing, but better naming and documenting can alleviate the problem. Alternatively, we could define a separate interface to act on the rearranged index and mark deletes. What do you think?
2. That’s true, whoever implements a document selector would have to identify deletes. I think you’re right - there’s no good reason to handle deletions as we’re rearranging. It can be a separate step, after the rearranged index has been produced.

zhaih reviewed (commented):
Thanks for making the change! However I don't see an advantage of putting this logic here because the interface is a pretty loose interface. In reality I think it is better to just let the document selector select all documents, then apply the deletion logic after all the index is rearranged?

stefanvodita commented:
Thank you for the review @zhaih ! Your feedback is much appreciated. I left a detailed response to your comment on the code. Let me know what you think.

zhaih commented:
I think the overall process in my mind would be:
1. Identify all deletes and save them
2. Create a document selector which ignores all deletes in the original index
3. Rearrange
4. Apply deletes

We can create a separate tool which do the step 1 and 4, or, which I think probably is more elegant, we can pass in another single DocumentSelector, which basically select which documents should be deleted afterwards. And we apply the deletions to the whole index after rearrange?

stefanvodita commented:
> we can pass in another single DocumentSelector, which basically select which documents should be deleted afterwards. And we apply the deletions to the whole index after rearrange?

I like the idea. I'll build a new revision following that line of thinking.

stefanvodita committed:
Apply deletes using DocumentSelector

stefanvodita commented:
The second revision comes with a lot more changes to support selecting deletes in the same fashion as segment content. I’ve reworked the tests to be more thorough, especially about deletes. I think the new tests cover all the cases that the old tests did, so I replaced the old ones.

This change is technically not backwards compatible. Not just because of changes to the rearrange API, but also because now we no longer make the deletes disappear from the rearranged index. They become live instead.

Once this PR goes through, I’ll start working on a change for luceneutil to use the new functionality.

mikemccand commented:
> This change is technically not backwards compatible. Not just because of changes to the rearrange API, but also because now we no longer make the deletes disappear from the rearranged index. They become live instead.

I think this is OK -- this is a new tool, not heavily used.  And keeping the deletes live is the purpose of this change :)

And `IndexRearranger` is already marked with `@lucene.experimental`, warning users that we might make non-backwards-compatible changes.

mikemccand commented on lucene/misc/src/java/org/apache/lucene/misc/index/IndexRearranger.java:
```diff
@@ -139,6 +203,47 @@ private static void addOneSegment(
+  private static void applyDeletesToOneSegment(
+      IndexWriter writer, CodecReader segmentReader, DocumentSelector selector) throws IOException {
+    Bits deletedDocs = selector.getFilteredDocs(segmentReader);
+    for (int i = 0; i < segmentReader.maxDoc(); ++i) {
```
Could we rename `i` to `docid`?

mikemccand commented on lucene/misc/src/java/org/apache/lucene/misc/index/IndexRearranger.java:
```diff
@@ -139,6 +203,47 @@ private static void addOneSegment(
+    for (int i = 0; i < segmentReader.maxDoc(); ++i) {
+      if (deletedDocs.get(i)) {
+        if (writer.tryDeleteDocument(segmentReader, i) == -1) {
+          throw new IllegalStateException("tryDeleteDocument failed and there's no plan B");
```
LOL.

Fortunately, as `tryDeleteDocument` is currently implemented today, it should never fail, since you have disabled merging in this writer.

mikemccand commented:
What happens if the delete selector deletes 100% of the documents in a segment?

`IndexWriter` would normally drop such segments ... does it do so in this case too?  Indeed, `testDeleteEverything` seems to confirm it does, great!

Maybe explain this caveat in the javadocs?  I.e. one cannot produce a segment geometry that includes segments with 100% deleted documents.

stefanvodita committed:
Clarify some javadoc, exception message, var name

stefanvodita commented:
@zhaih @mikemccand - is there anything more you'd like to see in this PR?

mikemccand commented:
Thanks for the reminder/ping @stefanvodita!  The latest iteration looks great to me -- I'll wait for a few days to see if @zhaih has more feedback and then merge.  It's exciting that Lucene can so accurately reproduce an index with precisely the same assignment of documents to segments, and now also with deletions.  Next up might be with doc-values updates too!  Maybe open a follow-on issue for this?

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/index/BinaryDocValueSelector.java:
```diff
@@ -32,26 +33,25 @@
 
   private final String field;
-  private final HashSet<String> keySet;
+  private final Set<String> keySet;
```
Since we're anyway rewriting this class quite a lot and is going to only put it under Lucene 10, let's change this `keySet` to `Set<BytesRef>`? Since the identifier doesn't need to be able to be converted to string. (That was my fault! Sorry!)

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/index/BinaryDocValueSelector.java:
```diff
@@ -60,18 +60,58 @@ public BitSet getFilteredLiveDocs(CodecReader reader) throws IOException {
+
         for (int i = 0; i < context.reader().maxDoc(); i++) {
-          if (liveDocs != null && liveDocs.get(i) == false) {
+          if (liveDocs == null || liveDocs.get(i) == true) {
```
move `liveDocs == null` before the loop so we can terminate earlier?

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/index/BinaryDocValueSelector.java:
```diff
@@ -60,18 +60,58 @@ public BitSet getFilteredLiveDocs(CodecReader reader) throws IOException {
+          if (binaryDocValues.advanceExact(i)) {
+            keySet.add(binaryDocValues.binaryValue().utf8ToString());
+          } else {
+            throw new AssertionError("Document don't have selected key");
```
We can put in the lucene doc id here to make it easier to debug?

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/index/IndexRearranger.java:
```diff
@@ -139,6 +206,48 @@ private static void addOneSegment(
+      return;
+    }
+
+    ExecutorService executor =
```
Let's reuse the executor?

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/index/IndexRearranger.java:
```diff
@@ -54,59 +68,83 @@
-   *     each selector correspond to one segment
+   * @param segmentSelectors specify what document is desired in the rearranged index segments, each
+   *     selector correspond to one segment
+   * @param deletedDocsSelector specify which documents are to be marked for deletion in the
```
Let's add a sentence indicating this selector needs to be thread safe?

zhaih reviewed (changes requested):
Thanks @stefanvodita! Sorry for reviewing lately.
The change looks great and I only have some small comments!

stefanvodita committed:
Make various improvements

Changes:
1. Fix test bug where documents wouldn't get deleted.
2. Strengthen the tests.
3. Store BinaryDocValues in the binary doc value selector as ByteRef
   instead of String.
4. Rename 'i' to 'docid' in some instances.
5. Make a live docs check earlier in the binary doc values selector.
6. Make AssertionError message more useful.
7. Use a single executor in IndexRearranger.
8. Specify that the deletes selector should be thread-safe.

stefanvodita commented:
Thank you for the ideas @zhaih , I’ve incorporated all of them. Your suggestion of using `ByteRef` in the keyset led to me finding a bug in the tests. The index creation methods didn’t delete any docs because I had stored the "delete" marker in a `BinaryDocValuesField`. That’s fixed now and the tests are more thorough with their checks.

@mikemccand , I don’t know how doc-values updates work. Why would they not be covered by this solution?

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/index/BinaryDocValueSelector.java:
```diff
@@ -30,52 +31,94 @@
+
+        for (int docid = 0; docid < context.reader().maxDoc(); docid++) {
+          if (binaryDocValues.advanceExact(docid)) {
+            keySet.add(new BytesRef(binaryDocValues.binaryValue().bytes.clone()));
```
use `BytesRef.deepCopyOf(binaryDocValues.binaryValue())` instead?
Directly copy underlying bytes array may cause some error when `offset` is not 0.

stefanvodita replied:
Oh, I see. Changed in latest commit.

zhaih commented on lucene/misc/src/java/org/apache/lucene/misc/index/BinaryDocValueSelector.java:
```diff
@@ -30,52 +31,94 @@
-            keySet.add(binaryDocValues.binaryValue().utf8ToString());
+
+          if (binaryDocValues.advanceExact(docid)) {
+            keySet.add(new BytesRef(binaryDocValues.binaryValue().bytes.clone()));
```
See comment below :)

stefanvodita committed:
Make ByteRef deep copies correctly

zhaih reviewed (approved):
LGTM, thank you @stefanvodita
